accepted/00000-replace-salt-toaster.md

```diff
@@ -0,0 +1,192 @@
+
+# Motivation
+[motivation]: #motivation
+
+<!-- - Why are we doing this? -->
+<!-- - What use cases does it support? -->
+<!-- - What is the expected outcome? -->
+
+<!-- Describe the problem you are trying to solve, and its constraints, without coupling them too closely to the solution you have in mind. If this RFC is not accepted, the motivation can be used to develop alternative solutions. -->
+
+Salt Toaster was developed a few years ago and filled a gap in the Salt world.
+Over the last years, maintaining the downstream `suse.tests` was too much effort
+-- so much that we didn't do it. Maintaining the docker images that Salt Toaster
+uses to run the upstream test suite against our patched Salt grew into a bigger
+and bigger task. We didn't keep up with upstream's changes to the test
+dependencies well and use a number of workarounds to still run a test suite.
+
+At the same time, Salt upstream improved their test suite a lot. They added new
+test types and now have a similar test framework to Salt Toaster. 
+
+The purpose of this change is to lessen our maintenance burden and to be good
+open-source citizens that collaborate with our upstream project when it comes to
+testing Salt. Additionally, we have knowledge about running a test suite within
+our team that we want to take advantage of.
+
+# Detailed design
+[design]: #detailed-design
+
+<!-- This is the bulk of the RFC. Explain the design in enough detail for somebody familiar with the product to understand, and for somebody familiar with the internals to implement. -->
+
+<!-- This section should cover architecture aspects and the rationale behind disruptive technical decisions (when applicable), as well as corner-cases and warnings. Whenever the new feature creates new user interactions, this section should include examples of how the feature will be used. -->
+
+## Overview
+
+The Salt upstream test suite is executed in virtual machines (VMs) that run the
+different operating systems we support. These VMs are set up with sumaform and
+salt-jenkins. Terracumber triggers terraform to create the VMs from within a
+Jenkins job and launches the test suite, connecting to each VM over SSH.
+
+## Provisioning Test VMs with Sumaform + Salt-Jenkins
+
+A new module "salt\_testenv" is added to sumaform. This module defines a new
+role, "salt\_testenv", and the default values needed to run all test suites, e.g.
+4GiB of RAM. Sumaform uses this role to apply the `golden-images-provision.sls`
+state file from the Salt-Jenkins repository.
+
+Up to now, all state files used by sumaform are maintained in the sumaform Git
+repository. For provisioning the salt\_testenv VMs, we don't want to maintain
+our own states that define which dependencies need to be installed on what
+operating system. Salt upstream already maintains such states for their own test
+infrastructure. We re-use their states by adding `salt-jenkins` as a Git
+submodule to sumaform and merge the salt state trees.
```

Comment:
What advantages and disadvantages you see with using submodules vs. consuming states from another repo (=defined path, for example)?

Reply:
A Git submodule is another Git repository, it's just a way of specifying the relationship. It's  possible to just `git clone` another Git repository instead, that way we would need to manually check out and we couldn't take advantage of a unified `git diff`. I don't really see a reason not to have the submodule relationship set up at this point.

The second question is more interesting to me, at least if I interpret it correctly: Is there a way to apply the states from another directory without merging them into `salt/`?
I don't know yet if it works and am curios to try a different approach: Could we inject this other directory as a second `file-root` into the `salt-call` execution that runs the provisioning and let Salt merge the state trees? 

I'm trying this out and see what it would take (if possible), I'll let you know when I got something.

Reply:
`--file-root` can only be used once, but it's possible to define multiple file roots in `/etc/salt/minion`. `sumaform` could generate such a config file (like it does for `/etc/salt/grains`), and we won't need to merge the trees ourselves.